File: 617.merge-two-binary-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode mergeTrees(TreeNode t1, TreeNode t2) {
        if (t1 == null)
            return t2;
        if (t2 == null)
            return t1;

        TreeNode temp = new TreeNode(t1.val + t2.val);
        temp.left = mergeTrees(t1.left, t2.left);
        temp.right = mergeTrees(t1.right, t2.right);

        return temp;
    }
}

File: 617.merge-two-binary-trees.js
/*
 * @lc app=leetcode id=617 lang=javascript
 *
 * [617] Merge Two Binary Trees
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} t1
 * @param {TreeNode} t2
 * @return {TreeNode}
 */
// var mergeTrees = function (t1, t2) {
//   if (!t1) return t2;
//   if (!t2) return t1;

//   t1.val += t2.val;

//   t1.left = mergeTrees(t1.left, t2.left);
//   t1.right = mergeTrees(t1.right, t2.right);
//   return t1;
// };


// BFS的做法
var mergeTrees = function (t1, t2) {
  if (!t1) return t2;
  let stack = new Array();
  stack.push([t1, t2]);

  while (stack.length) {
    let nodeArr = stack.shift();
    if (nodeArr[1] == null) {
      continue;
    };
    nodeArr[0].val += nodeArr[1].val;
    if (nodeArr[0].left == null) {
      nodeArr[0].left = nodeArr[1].left;
    } else {
      stack.push([nodeArr[0].left, nodeArr[1].left])
    }

    if (nodeArr[0].right == null) {
      nodeArr[0].right = nodeArr[1].right;
    } else {
      stack.push([nodeArr[0].right, nodeArr[1].right]);
    }
  }

  return t1;
};

